geo/src/algorithm/relate/geomgraph/index/simple_edge_set_intersector.rs

```diff
@@ -0,0 +1,59 @@
+use super::super::Edge;
+use super::{EdgeSetIntersector, SegmentIntersector};
+use crate::GeoFloat;
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+pub(crate) struct SimpleEdgeSetIntersector;
+
+impl SimpleEdgeSetIntersector {
+    pub fn new() -> Self {
+        SimpleEdgeSetIntersector
+    }
+
+    fn compute_intersects<F: GeoFloat>(
+        &mut self,
+        edge0: &Rc<RefCell<Edge<F>>>,
+        edge1: &Rc<RefCell<Edge<F>>>,
+        segment_intersector: &mut SegmentIntersector<F>,
+    ) {
+        let edge0_coords_len = edge0.borrow().coords().len() - 1;
+        let edge1_coords_len = edge1.borrow().coords().len() - 1;
+        for i0 in 0..edge0_coords_len {
+            for i1 in 0..edge1_coords_len {
+                segment_intersector.add_intersections(edge0, i0, edge1, i1);
+            }
+        }
+    }
+}
+
+impl<F: GeoFloat> EdgeSetIntersector<F> for SimpleEdgeSetIntersector {
+    fn compute_intersections_within_set(
+        &mut self,
+        edges: &[Rc<RefCell<Edge<F>>>],
+        check_for_self_intersecting_edges: bool,
+        segment_intersector: &mut SegmentIntersector<F>,
+    ) {
+        for edge0 in edges.iter() {
+            for edge1 in edges.iter() {
+                if check_for_self_intersecting_edges || edge0.as_ptr() != edge1.as_ptr() {
+                    self.compute_intersects(edge0, edge1, segment_intersector);
```

Comment:
I believe there were two reasons for the unfortunate use of `Rc<RefCell<>>`.

The lesser one is that RelateOperation grabs a duplicated reference to `Edge` at some point. I think we could get rid of that usage by storing an arena-style index-of-edge instead of a pointer-to-edge.

The bigger reason is here: `compute_intersection` calls `segment_intersector.add_intersection` which, as well as *computing* the intersections, also mutates the edge to record those intersections **and it might be intersecting an edge with itself**.

This all makes for some pretty unremarkable Java code, but rust makes it painfully clear that this is less-than-nice.

I think it would be *possible* to remove the `Rc<RefCell<>>` stuff with more work, but I decided to leave it as is for now.

FWIW I did some profiling and the RefCell/Rc stuff does not significantly contribute to runtime.